src/widgets/models/FormulaModel.js

```diff
@@ -1,22 +1,35 @@
-import { executeSQL} from '../../api';
-import { filtersToSQL, viewportToSQL } from '../../api/FilterQueryBuilder';
+import { executeSQL } from '../../api';
+import { filtersToSQL } from '../../api/FilterQueryBuilder';
+import { applyFilter } from '../../api/Filter';
+import { aggregationFunctions } from '../operations/aggregation/values';
 
 export const getFormula = async (props) => {
-  const { data, credentials, operation, column, filters, viewport, opts } = props;
+  const { data, credentials, operation, column, filters, opts, viewportFilter, viewportFeatures, type } = props;
 
   if (Array.isArray(data)) {
     throw new Error('Array is not a valid type to get categories');
   }
 
-  let query =
-    (viewport && `SELECT * FROM (${data})  as q WHERE ${viewportToSQL(viewport)}`) ||
-    data;
+  if (type === 'bq' && !viewportFilter) {
+    throw new Error('Formula Widget error: BigQuery layers needs "viewportFilter" prop set to true.');
+  }
+
+  // It's an await because we probably will move this calculation need to a webworker
+  if (viewportFilter) {
+    const targetOperation = aggregationFunctions[operation];
 
-  query = `
-    SELECT ${operation}(${column}) as value
-    FROM (${query}) as q
-    ${filtersToSQL(filters)}
-  `;
+    const features = viewportFeatures || [];
+    const filteredFeatures = features.filter(applyFilter({ filters }));
+    return await [{ value: targetOperation(filteredFeatures, column) }];
```

Comment:
Why do you use return await + array here?

Reply:
We were planning to move this to a web worker, for now it's not necessary, removing..